### PR TITLE
feat(defi): consistent handling of insufficient ETH for approval gas

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR:

- consistently displays the `Not enough %{assetSymbol} to cover gas` info for DeFi opportunities approval step
- displays the `There will be separate gas fee to deposit/withdraw` alert for DeFi opportunities approval step in case there isn't enough native token (ETH in all touched files) in the wallet to cover gas
- disables the confirm button in case of insufficient balance for gas

This PR does **not** have multi-account guarantees as we're using the `selectPortfolioCryptoHumanBalanceByAssetId` selector, but since it's already boilerplaty as is, expect a follow-up for multi-account.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- tackles https://github.com/shapeshift/web/issues/2945

## Testing

- For all DeFi opportunities types, the confirm button at deposit/withdraw approve steps should be disabled in case of not enough ETH for gas
- For all DeFi opportunities types, an alert should be shown (see screenshots) in case of not enough ETH for gas
- Whether or not there is enough ETH for gas, `Not enough %{assetSymbol} to cover gas` should be consistently displayed

### Engineering

- See top-level notes

### Operations

- See top-level notes

## Screenshots (if applicable)

image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #2944** <a href="https://app.graphite.dev/github/pr/shapeshift/web/2944" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/2944?utm_source=stack-comment).

File: src/features/defi/providers/foxy/components/FoxyManager/Deposit/components/Approve.tsx [hunk 1]
```diff
@@ -188,12 +200,22 @@ export const Approve: React.FC<ApproveProps> = ({ accountId, onNext }) => {
       loadingText={translate('common.approve')}
       learnMoreLink='https://shapeshift.zendesk.com/hc/en-us/articles/360018501700'
       preFooter={
-        <Alert status='info' borderRadius='lg' color='blue.500'>
-          <FaGasPump />
-          <AlertDescription textAlign='left' ml={3} color={alertText}>
-            {translate('modals.approve.depositFee')}
-          </AlertDescription>
-        </Alert>
+        <>
```
Comment (gomesalexandre) [inline src/features/defi/providers/foxy/components/FoxyManager/Deposit/components/Approve.tsx:203]: Memo

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/components/Approve.tsx [hunk 2]
```diff
@@ -193,12 +206,23 @@ export const Approve: React.FC<StepComponentProps & { accountId: Nullable<Accoun
       loadingText={translate('common.approve')}
       learnMoreLink='https://shapeshift.zendesk.com/hc/en-us/articles/360018501700'
       preFooter={
-        <Alert status='info' borderRadius='lg' color='blue.500'>
-          <FaGasPump />
-          <AlertDescription textAlign='left' ml={3} color={alertText}>
-            {translate('modals.withdraw.withdrawFee')}
-          </AlertDescription>
-        </Alert>
+        <>
```
Comment (gomesalexandre) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/components/Approve.tsx:209]: Memo

File: src/features/defi/components/Approve/Approve.tsx [hunk 3]
```diff
@@ -36,20 +36,21 @@ type ApproveProps = {
 
 export const Approve = ({
   asset,
+  contractAddress,
   cryptoEstimatedGasFee,
+  disabled,
```
Comment (gomesalexandre) [inline src/features/defi/components/Approve/Approve.tsx:41]: OCD reorder, just adds `disabled`

**Review by 0xApotheosis — APPROVED** [review]
This works as expected 👌 

Screen Shot 2022-10-10 at 7 05 35 pm

Pre-stamping, though I have one comment on the duplication of the implementation that I'd prefer to see addressed before merging:

We create 8 `preFooter` elements throughout the codebase now, and they only differ by their translation string.

Can we consolidate these, perhaps into a pure function that takes a translation key and `hasEnoughBalanceForGas`, or a custom hook that does the `hasEnoughBalanceForGas` logic itself.

**Comment by 0xApotheosis** [comment]
Since you are away gomes I made a follow-up ticket to tackle the comment above (https://github.com/shapeshift/web/issues/3014), so I'll now merge this one.

**Comment by tshifty** [comment]
This works for account #0 but does not work for any new accounts added for ETH

**Pull request merged**